### ISSUE TITLE
`set_rectangle` adds unnecessary padding even when batch shape is already divisible by stride

### ISSUE THREAD

**Issue by a17sol** [ultralytics/ultralytics#24388]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I noticed that when using `rect=True` for validation, batches have an extra padding even if the shape is already divisible by stride.
The `build_yolo_dataset` function creates `YOLODataset` with the argument `pad=0.0 if mode == "train" else 0.5`, and `set_rectangle` computes batch shapes as follows:
```
self.batch_shapes = np.ceil(np.array(shapes) * self.imgsz / self.stride + self.pad).astype(int) * self.stride
```
That means that the shape is always rounded up to the next multiple of stride.
For example, a single 128×128 image with `imgsz=640` and `stride=32` gets `batch_shape=672` instead of 640, adding 32px of unnecessary padding.
This creates an inconsistency with `predict`, which pads only to the nearest multiple of stride and produces no padding in this case.
Is this `pad=0.5` intentional? If so, what is the reasoning behind it? If not, would it make sense to adjust the formula so that padding is only added when actually needed to reach the next stride multiple?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @a17sol, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue triaged quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Since your report discusses a possible inconsistency in `rect=True` validation padding behavior, a small standalone repro would be especially helpful 🔍
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
You're right — with val/test `pad=0.5`, [`set_rectangle()`](https://docs.ultralytics.com/reference/data/base/#ultralytics.data.base.BaseDataset.set_rectangle) will round an already stride-aligned batch from `640` to `672`, while [`LetterBox`](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.LetterBox) in `predict()` only pads when needed. This is intentional because validation runs at a higher `imgsz`, using the extra compute that is gained by using rectangular input.

**Comment by Y-T-G** [comment]
Hello  Thanks for raising this.

Yes, this is intentional

**Comment by a17sol** [comment]
Thank you for the answers! 
@Y-T-G I didn't fully understand the explanation you added by editing the Glenn's answer. Could you elaborate on why the extra padding improves validation quality? Intuitively it seems like padding with empty pixels shouldn't help the model detect objects better compared to simply using the exact stride-aligned shape. Is it about giving more context around objects near image borders, or is there some other reasoning?

**Comment by Y-T-G** [comment]
It's mostly empirical. You can compare the results with and without padding to see the difference. It has higher mAP with padding. It's not obvious why. Could be that the objects at the edges benefit from it.

**Comment by Y-T-G** [comment]
The original intention probably was to upscale the image to a larger imgsz, not just pad. But the dataloader downscales all images to imgsz. And then Letterbox doesn't upscale to larger imgsz obtained with the `pad` because `scaleUp` is disabled during validation, which results in blank padding. So it's a bug that somehow helped and remained a feature.

**Comment by a17sol** [comment]
If models empirically work better with this padding in `val`, wouldn't it make sense to add the same padding to `predict` as well?

**Comment by Y-T-G** [comment]
It works better with padding plus higher imgsz. Not just one of them alone. To add it to predict, it would require running prediction on an imgsz larger than what the user specified, which would cause issues because that's not what the user wants.

**Comment by a17sol** [comment]
Thanks for the explanation! But doesn't this mean that `val` metrics don't actually reflect real-world inference performance? If `val` uses a different preprocessing than `predict`, then a higher mAP in `val` doesn't necessarily mean the model will perform better in practice. The goal of validation is to estimate how well the model works in deployment, not to maximize numbers on a benchmark. Wouldn't it be more honest to have `val` and `predict` use identical preprocessing, even if that means slightly lower `val` metrics?

**Comment by Y-T-G** [comment]
> Wouldn't it be more honest to have val and predict use identical preprocessing, even if that means slightly lower val metrics?

You can obtain that by setting `rect=False` during validation which doesn't increase `imgsz` and uses square padding, which is the typical deployment scenario.

> If val uses a different preprocessing than predict, then a higher mAP in val doesn't necessarily mean the model will perform better in practice.

It generally should. If the model improves, then it should relatively improve in both cases. It's unlikely for it to improve in `rect=True` validation, while not improving in `rect=False`. So a higher validation mAP is a sign of improvement as long as you're comparing between same modes, either with `rect` or without `rect`.
